libopae/src/error.c

```diff
@@ -0,0 +1,328 @@
+			return FPGA_OK;
+		}
+		i++;
+		p = p->next;
+	}
+
+	FPGA_MSG("error info %d not found", error_num);
+	return FPGA_NOT_FOUND;
+}
+
+/* files and directories to ignore when looking for errors */
+const char *errors_exclude[] = {
+	"revision",
+	"uevent",
+	"power",
+	"clear"
+};
+#define NUM_ERRORS_EXCLUDE (sizeof(errors_exclude) / sizeof(char *))
+
+/* files that can be cleared by writing their value to them */
+const char *errors_clearable[] = {
+	"pcie0_errors",
+	"pcie1_errors",
+	"warning_errors",
+	"inject_error"
+};
+#define NUM_ERRORS_CLEARABLE (sizeof(errors_clearable) / sizeof(char *))
+
+/* Walks the given directory and adds error entries to `list`.
+ * This function is called during enumeration when adding tokens to
+ * the global tokens list. When tokens are cloned, their error
+ * lists are only shallowly copied (which works because errors of
+ * a token never change).
+ * returns the number of error entries added to `list` */
+uint32_t __FIXME_MAKE_VISIBLE__
+build_error_list(const char *path, struct error_list **list)
+{
+	struct dirent *de;
+	DIR *dir;
+	struct stat st;
+	char basedir[FILENAME_MAX];
+	int len = strlen(path);
+	uint32_t n = 0;
+	unsigned int i;
+	struct error_list **el = list;
+
+	if (len+1 > FILENAME_MAX) {
+		FPGA_MSG("path too long");
+		return 0;
+	}
+
+	strncpy(basedir, path, FILENAME_MAX-1);
+	basedir[len++] = '/';
+
+	dir = opendir(path);
+	if (!dir) {
+		FPGA_MSG("unable to open %s", path);
+		return 0;
+	}
+
+	while ((de = readdir(dir))) {
+		// skip hidden ('.*') files (includes "." and "..")
+		if (de->d_name[0] == '.')
+			continue;
+
+		// skip names on blacklist
+		for (i = 0; i < NUM_ERRORS_EXCLUDE; i++) {
+			if (strcmp(de->d_name, errors_exclude[i]) == 0) {  // FIXME: SAFE?
+				break;
+			}
+		}
+		if (i < NUM_ERRORS_EXCLUDE)
+			continue;
+
+		// build absolute path
+		strncpy(basedir + len, de->d_name, FILENAME_MAX - len);
+
+		// try accessing file/dir
+		if (lstat(basedir, &st) == -1) {
+			FPGA_MSG("can't stat %s", basedir);
+			continue;
+		}
+
+		// skip symlinks
+		if (S_ISLNK(st.st_mode))
+			continue;
+
+		// recursively dive into subdirectories
+		if (S_ISDIR(st.st_mode)) {
+			n += build_error_list(basedir, el);
+			continue;
+		}
+
+		// not blacklisted, not hidden, accessible, no symlink, no dir -> count and append it!
+		n++;
+		if (!el)	// no list
+			continue;
+
+		// append error info to list
+		struct error_list *new_entry = malloc(sizeof(struct error_list));
+		if (!new_entry) {
+			FPGA_MSG("can't allocate memory");
+			n--;
+			break;
+		}
+		strncpy_s(new_entry->info.name, FPGA_ERROR_NAME_MAX, de->d_name, FILENAME_MAX);
+		strncpy_s(new_entry->error_file, SYSFS_PATH_MAX, basedir, FILENAME_MAX);
+		new_entry->next = NULL;
+		// Errors can be cleared:
+		//   * if the name is "errors" and there is a file called "clear" (generic case), OR
+		//   * if the name is in the "errors_clearable" table
+		new_entry->info.can_clear = false;
```

Comment:
Should the routine to check if an error file is clearable be refactored into its own function?

Reply:
Good idea. I'll look into it.

Reply:
Ok, it can be a TODO...